### PR TITLE
Dependency rules

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
This is what I described in https://github.com/ansible-community/community-topics/issues/94

Let's polish this a bit, and if everyone seems happy let's vote on the result.

##### ISSUE TYPE
- Docs Pull Request
- Feature Pull Request

##### COMPONENT NAME
inclusion criteria

### CODE REVIEW

File: collection_requirements.rst
Comment (Andersson007) [inline collection_requirements.rst:403]:
```suggestion
  #. They have a loose dependency on one (or more) major versions of other collections included in Ansible. For example, ``ansible.netcommon: >=1.0.0``, or ``ansible.netcommon: >=2.0.0, <3.0.0``. In case the collection depended on releases a new major version outside of this version range that will be included in the next major Ansible release, the dependent collection will be removed from the next major Ansible release. The cut-off date for this is feature freeze.
```

File: collection_requirements.rst
Comment (Andersson007) [inline collection_requirements.rst:409]:
```suggestion
#. ``community.foo 1.2.0`` has a dependency on ``community.bar >= 1.0.0, < 1.3.0``.
```

File: collection_requirements.rst
Comment (Andersson007) [inline collection_requirements.rst:418]:
```suggestion
   * Now ``community.bar`` creates a new release ``1.3.0``. When ``community.foo`` does not create a new release with a relaxed dependency, we have to include ``community.bar 1.2.x`` in the next Ansible release despite ``1.3.0`` being available.
   * If ``community.foo`` does not relax its dependency on ``community.bar`` for some time, ``community.foo`` will be removed from the next Ansible major release.
   * Unfortunately ``community.bar`` has to stay at ``1.2.x`` until either community.bar is removed (in the next major release), or loosens its requirements so that newer ``community.foo 1.y.z`` releases can be included.

#. ``community.foonetwork`` depends on ``ansible.netcommon >= 2.0.0, <3.0.0``.

   * ``ansible.netcommon 4.0.0`` is released during this major Ansible release cycle.
   * ``community.foonetwork`` either releases a new version before feature freeze of the next major Ansible release that allows depending on all ``ansible.netcommon 4.x.y`` releases, or it will be removed from the next major Ansible release.
```

File: collection_requirements.rst
Comment (Andersson007) [reply collection_requirements.rst:403]: also what does `a loose dependency` mean? I googled a bit but failed to find something meaningful to me. Should we elaborate on it a bit more if the word `loose` is important here? If not, i think we should re-phrase it somehow.

**Review by Andersson007 — COMMENTED** [review]
@felixfontein generally LGTM, a couple of minor things from me, thanks for doing this!

File: collection_requirements.rst
Comment (felixfontein) [reply collection_requirements.rst]: Yep, "dependend" doesn't exist. I've rewritten this sentence and made another small change. Can you check whether this is better now?

Comment (Andersson007) [reply collection_requirements.rst]: I would suggest using:
1. `dependency` instead of `depended on` as there are a lot of `dependency|ies` are used across the paragraph and `depended on` only once OR
2. change `dependency|ies` to `collections depended on` everywhere (though the option 1 feels a bit better to me.

Generally the combination of `depended on` and `dependent collection` confuses me a lot:) Makes me focusing each time to figure out which means what:) Can be fair for other readers, i guess

Comment (felixfontein) [reply collection_requirements.rst]: The problem is that `dependency` is the relation between two collections, while `the collection dependend on` is one of the two collections involved. I skimmed over all uses of `dependenc*`, and I think its used correctly everywhere to indicate the actual dependency (as the relation), and not the dependend on collection.

Comment (Andersson007) [reply collection_requirements.rst]: OK but it confuses me anyway, though maybe it's my personal feature

**Review by Andersson007 — APPROVED** [review]

**Comment by felixfontein** [comment]
Hmm, looks like I lost commit access to this repository, so I cannot even update my own branch anymore :-( CC @gundalow @Andersson007

File: collection_requirements.rst
Comment (russoz) [inline collection_requirements.rst:418]: This should be determined automatically, and there should be some sort of a state-file holding this info between releases.

File: collection_requirements.rst
Comment (russoz) [inline collection_requirements.rst:388]: Could you please expand what you mean by `under development`? Would that be `<= 1.0.0`?

Comment (felixfontein) [reply collection_requirements.rst:388]: Yes. This is basically about two dependent collections simultaneously reaching the 1.0.0 goal while producing versions that satisfy our rules.

File: collection_requirements.rst
Comment (felixfontein) [reply collection_requirements.rst:418]: This sounds pretty complicated for something that hopefully never happens, and if it does can be handled with issues and already preparing removal from the upcoming major version (as its `ansible.in` already exists in a new directory).

File: collection_requirements.rst
Comment (russoz) [reply collection_requirements.rst:388]: Wouldn't it make sense to bar not only the collections under dev but also the ones that _depend on_ collections under dev? I mean, if the purpose of the rule is reliability/stability, I suppose that would go a long way.

Comment (felixfontein) [reply collection_requirements.rst:388]: We only accept collections that released version 1.0.0 for inclusion anyway. Also note that this part of this PR has been in the rules for years, it only moved to a new place in this document.

File: collection_requirements.rst
Comment (felixfontein) [reply collection_requirements.rst]: Thanks!

File: collection_requirements.rst
Comment (Dule-mart) [inline collection_requirements.rst:118]: Collection dependencies must meet a set of rules. See the section on `Collection Dependencies <collection_dependencies_>`_ for this.

Comment (Dule-mart) [reply collection_requirements.rst:118]: I think it is easier to read the word "meet" than "satisfy" reason is that the word "meet" is a common word.

File: collection_requirements.rst
Comment (Dule-mart) [reply collection_requirements.rst:388]: I'm concerned with the use of the word "under development", what if that could be changed to "under developed"

Let's assume this - * When creating new collections where collection dependencies are also under developed,
What do you think?

Comment (felixfontein) [reply collection_requirements.rst:388]: I'm not a native English speaker, but "under developed" seems to mean something completely different to me. ("in development" would be an alternative - https://english.stackexchange.com/questions/238497/which-expression-is-correct-in-development-or-under-development)

File: collection_requirements.rst
Comment (felixfontein) [reply collection_requirements.rst:118]: If you add this as a suggestion I can merge it with attribution to you.

Comment (Dule-mart) [reply collection_requirements.rst:118]: Ok

File: collection_requirements.rst
Comment (Dule-mart) [reply collection_requirements.rst:388]: Oooh, interesting I just learned that now. Thanks

File: collection_requirements.rst
Comment (Dule-mart) [inline collection_requirements.rst:118]: Collection dependencies must meet a set of rules. See the section on Collection Dependencies <collection_dependencies_>_ for this.

**Review by Dule-mart — CHANGES_REQUESTED** [review]

File: collection_requirements.rst
Comment (felixfontein) [reply collection_requirements.rst]: Thanks!

File: collection_requirements.rst
Comment (jamescassell) [inline collection_requirements.rst:397]: to avoid blocking the collection being depended upon

Comment (felixfontein) [reply collection_requirements.rst:397]:
```suggestion
* If a collection has a too strict dependency for a longer time, and forces another collection to be hold back, that collection will be removed from the next major Ansible release. What "longer time" means depends on when the next Ansible major release happens. If a dependent collection prevents a new major version of a collection it depends on to be included in the next major Ansible release, the dependent collection will be removed from that major release to avoid blocking the collection being dependend upon.
```

**Comment by felixfontein** [comment]
The new rules have been accepted (https://github.com/ansible-community/community-topics/issues/94#issuecomment-1203722898) by SC vote. Any last comments before merging?

File: collection_requirements.rst
Comment (samccann) [inline collection_requirements.rst:397]:
```suggestion
* If a collection has a too strict dependency for a longer time, and forces another collection depended on to be held back, that collection will be removed from the next major Ansible release. What "longer time" means depends on when the next Ansible major release happens. If a dependent collection prevents a new major version of a collection it depends on to be included in the next major Ansible release, the dependent collection will be removed from that major release to avoid blocking the collection being depended on.
```

File: collection_requirements.rst
Comment (samccann) [inline collection_requirements.rst:413]:
```suggestion
   * Unfortunately ``community.bar`` has to stay at ``1.2.x`` until either ``community.foo`` is removed (in the next major release), or loosens its requirements so that newer ``community.bar 1.3.z`` releases can be included.
```

Comment (samccann) [reply collection_requirements.rst:413]: not sure if I understand this correctly, but that's what I'm basing this change on above ^^

**Comment by felixfontein** [comment]
@Andersson007 @russoz @Dule-mart @jamescassell @samccann thanks a lot for reviewing this and suggesting improvements!

**Comment by Dule-mart** [comment]
Is all a pleasure doing this with everyone

**Pull request merged**
